dipy/denoise/localpca.py

```diff
@@ -0,0 +1,114 @@
+            for j in range(patch_radius, arr.shape[1] - patch_radius, 1):
+                for i in range(patch_radius, arr.shape[0] - patch_radius, 1):
+
+                    X = np.zeros(
+                        (patch_size * patch_size * patch_size, arr.shape[3]))
+                    M = np.zeros(arr.shape[3])
+
+                    temp = arr[i - patch_radius: i + patch_radius + 1,
+                               j - patch_radius: j + patch_radius + 1,
+                               k - patch_radius: k + patch_radius + 1, :]
+                    X = temp.reshape(
+                        patch_size *
+                        patch_size *
+                        patch_size,
+                        arr.shape[3])
+                    # compute the mean and normalize
+                    M = np.mean(X, axis=0)
+                    X = X - np.array([M, ] * X.shape[0],
+                                     dtype=np.float64)
+
+                    # Compute the covariance matrix C = X_transpose X
+                    C = np.transpose(X).dot(X)
+                    C = C / X.shape[0]
+
+                    [d, W] = np.linalg.eigh(C)
+
+                    d[d < tou[i, j, k, :]] = 0
+
+                    W_hat = W * 0
+                    W_hat[:, d > 0] = W[:, d > 0]
+                    Y = X.dot(W_hat)
+                    X_est = Y.dot(np.transpose(W_hat))
+
+                    # theta value
+                    temp = X_est + \
+                        np.array([M, ] * X_est.shape[0], dtype=np.float64)
+                    temp = temp.reshape(
+                        patch_size, patch_size, patch_size, arr.shape[3])
+                    # Also update the estimate matrix which is X_est * theta
+
+                    theta[i - patch_radius: i + patch_radius + 1,
+                          j - patch_radius: j + patch_radius + 1,
+                          k - patch_radius: k + patch_radius + 1,
+                          :] = theta[i - patch_radius: i + patch_radius + 1,
+                                     j - patch_radius: j + patch_radius + 1,
+                                     k - patch_radius: k + patch_radius + 1,
+                                     :] + 1.0 / (1.0 + np.linalg.norm(d,
+                                                                      ord=0))
+
+                    thetax[i - patch_radius: i + patch_radius + 1,
+                           j - patch_radius: j + patch_radius + 1,
+                           k - patch_radius: k + patch_radius + 1,
+                           :] = thetax[i - patch_radius: i + patch_radius + 1,
+                                       j - patch_radius: j + patch_radius + 1,
+                                       k - patch_radius: k + patch_radius + 1,
+                                       :] + temp / (1 + np.linalg.norm(d,
+                                                                       ord=0))
+
+        denoised_arr = thetax / theta
```

Comment:
Ah yes, also this correction is most noticeable at low SNR, as with a high enough SNR a gaussian approximation is enough, so it could be why you don't see too much difference if you have a high SNR. Here is a graph form the original paper that proposed this correction
![capture d ecran de 2016-08-16 19-55-14](https://cloud.githubusercontent.com/assets/3030760/17709926/66782f6a-63eb-11e6-99df-a4a514fe8675.png)

It is equation 4 from the lpca paper, and closely related to eq. 5. As you can see, above SNR 4 the correction factor is almost 1, and thus does not produce that much of a difference in the Rician noise case. It is more pronounced for other cases, but at low SNR it should give you different results though.

Figure 1 (or equation 8 in this ref) is from Koay, C. G., & Basser, P. J. (2006). Analytically exact correction scheme for signal extraction from noisy magnitude MR signals. Journal of Magnetic Resonance, 179(2), 317–322. http://doi.org/10.1016/j.jmr.2006.01.016